### DISCUSSION TITLE
Ability to discover / register bb tasks in monorepo and call them from root

### DISCUSSION THREAD

**Discussion by ieugen** [babashka/babashka#1044] [Ideas]
Posted on Clojurians Slack https://clojurians-log.clojureverse.org/babashka/2021-10-27 .

one issue I have with bb tasks and I hope it gets improved is that bb.edn files are spread across the monorepo and people don't know about the tasks available to them.
Drawing inspiration from gradle (that also has tasks) - it has the concept of sub-projects and it can include other builds thus making all tasks accessible from the root via a path  - :my-subproject:my-task ( uses semicolon for path separator).
Can babasha tasks support something like this ?
Should I open an issue ?
I woule like to be in monorepo root, do bb tasks  and get tasks that are in other subprojects.
Also I should be able to call those tasks from the root and they should be executed properly (as from their own dir).
Making the tasks available could be done via am include/require mechanism or a dir scan with filters (path is ~ namespace).
( I would pick include and maybe scan could be implemented in lib that produces a list of require bb files)
wdyt?

**Comment by borkdude** [comment]
One possible solution to this problem is to define a top level `bb.edn` and define a task for each sub-repository:

``` clojure
{:tasks {:requires ([clojure.string :as str])
         app {:doc "Without arguments, prints tasks in app - with arguments, invokes task in app dir"
              :task (if (empty? *command-line-args*)
                      (shell {:dir "app"} "bb tasks")
                      (shell {:dir "app"} (str "bb run " (str/join " " *command-line-args*))))}}}
```

Then when you run `bb app`, bb will list all tasks from the `app` directory, while invoking `bb app dev` will invoke `bb dev` in the app subdirectory. 

```
bb app ;; prints app tasks
bb app dev
bb app build
bb app publish-image
```

If you have many sub-repositories, of course you can also programmatically create the root bb.edn.

**Reply by mkarp** [reply]
@borkdude Would that mean that one would always have to run `bb` from the monorepo root?

**Reply by mkarp** [reply]
How would cross-project task dependencies look like with this solution?

**Reply by borkdude** [reply]
@mkarp 

Note that this solution already works today with babashka as is. As before, babashka looks at the `bb.edn` in the local directory, so which tasks are available depends on that file. You don't have to always execute from the root, you can also enter the `app` dir and then execute `bb dev`, etc. 

The second question: same as always. Just declare them in `:deps` and `:paths`.

**Reply by mkarp** [reply]
That's cool! I'll try to give it a shot. Thank you @borkdude!

**Comment by borkdude** [comment]
A few more thoughts here:

- Generate bb.edn using a script https://github.com/babashka/babashka/issues/1219#issuecomment-1128017087
- Auto-discover tasks in sub-directories https://github.com/babashka/babashka/issues/1219#issuecomment-1128031344 which can be invoked with `bb subdir/task`

**Reply by ieugen** [reply]
I'm experimenting with a version like this:
I wrote a function to delegate task execution to a directory. 

```clojure
{:min-bb-version "0.7.4"
 :tasks {:requires ([clojure.string :as str])
         :init (def bb-delegate (fn [dir]
                                  (if (empty? *command-line-args*)
                                    (shell {:dir dir} "bb tasks")
                                    (shell {:dir dir}
                                           (str "bb run " (str/join " " *command-line-args*))))))
         app {:doc "Without arguments, prints tasks in app - with arguments, invokes task in app dir"
              :task (bb-delegate "projects/app")}
         annotator {:doc "Without arguments, prints tasks in app - with arguments, invokes task in app dir"
                    :task (bb-delegate "projects/annotator")}}}
```

**Reply by borkdude** [reply]
Looks good!

**Reply by lispyclouds** [reply]
Moving over some of the thoughts from the #babashka-sci-dev Slack [thread](https://clojurians.slack.com/archives/C02FBBU61A9/p1652728131659869):
- Add a new key in the bb.edn called `:subprojects` That is a vector of strings showing the dirs where bb.edns are there and can be used when listing and/or running tasks. The pwd for the subtasks would be relative to the bb.edn they're defined in.
- All the subprojects have their own bb.edn and bb recursively discovers each of the projects top down and includes those dirs when listing/running tasks. Dirs not having a bb.edn would be ignored along with its children. Complex, maybe not worth it.
- Similar to the `subprojects` idea, we could use the vocabulary used in Rust's Cargo [workspaces](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html). Instead of calling it `subprojects`, maybe call it `workspaces` and `members` if more granularity is needed.

**Comment by just-sultanov** [comment]
This is an example of configuring a mono repository using [babashka](https://github.com/babashka/babashka) and [direnv](https://direnv.net/)

[Link to the example repository](https://github.com/just-sultanov/bb-modules)

**Comment by yenda** [comment]
@borkdude I like this hallucinated answer from Claude on the matter:

```
Babashka provides a :tasks-in key specifically for this purpose. This allows you to include task definitions from other EDN files.
Here's how to set it up:

Create separate EDN files for different task collections:

project/
├── bb.edn             # Main task file
├── build-tasks.edn    # Build-related tasks
└── deploy-tasks.edn   # Deployment-related tasks

In your main bb.edn file, use the :tasks-in key to include these files:

{:tasks {}  ; Can be empty or contain common tasks
 
 :tasks-in ["build-tasks.edn" 
            "deploy-tasks.edn"]}
```
